examples/direct_fidelity_estimation.py

```diff
@@ -82,35 +88,93 @@ def direct_fidelity_estimation(circuit: cirq.Circuit, qubits: List[cirq.Qid],
     inv_sum_p = 1 / sum(p)
     norm_p = [x * inv_sum_p for x in p]
 
-    fidelity = 0.0
-    for _ in range(n_trials):
-        # Randomly sample as per probability.
-        i = np.random.choice(range(4**n_qubits), p=norm_p)
+    simulator = cirq.DensityMatrixSimulator(noise=noise)
+    if sampled:
+        fidelity = 0.0
+        for _ in range(n_trials):
+            # Randomly sample as per probability.
+            i = np.random.choice(len(pauli_traces), p=norm_p)
+
+            Pr_i = pauli_traces[i]['Pr_i']
+            P_i = pauli_traces[i]['P_i']
+            rho_i = pauli_traces[i]['rho_i']
+
+            pauli_string = cirq.PauliString(dict(zip(qubits, P_i)))
+            qubit_map = dict(zip(qubits, range(n_qubits)))
+
+            p = cirq.PauliSumCollector(circuit=circuit,
+                                       observable=pauli_string,
+                                       samples_per_term=samples_per_term)
+
+            completion = p.collect_async(sampler=simulator)
+            cirq.testing.assert_asyncio_will_have_result(completion, None)
+
+            sigma_i = p.estimated_energy()
+            assert np.isclose(sigma_i.imag, 0.0, atol=1e-6)
+            sigma_i = sigma_i.real
+
+            fidelity += Pr_i * sigma_i / rho_i
+
+        return fidelity
+    else:
+        # sigma in https://arxiv.org/pdf/1104.3835.pdf
+        density_matrix = cast(cirq.DensityMatrixTrialResult,
+                              simulator.simulate(circuit)).final_density_matrix
+
+        fidelity = 0.0
+        for _ in range(n_trials):
```

Comment:
there's a decent amount of shared code in these two branches. Is there any way to factor it out?

Reply:
Refactored the code to have a separate function.